[Java] Document references (#2153)
cap-js/docs: java/working-with-cql/query-api.md

### Entity References {#entity-refs}

> Entity references specify entity sets and define the target entity set of a [CQL](../../cds/cql) statement or be an argument of event handler. 

Reviewer: ```suggestion
Entity references specify entity sets and can define the target entity set of a [CQL](../../cds/cql) statement or serve as an argument for event handlers.
```

> Reference consists of _segments_ that define the path from the entity's root to the certain part of it. Each segment has the _identifier_ with the name of the entity or an element and an optional filter _predicate_. These predicates might include other references.

Reviewer: ```suggestion
A reference consists of _segments_ that define the path from the entity's root to the certain part of it. Each segment has the _identifier_ with the name of the entity or an element and an optional filter _predicate_. These predicates might include other references.
```

References can be represented in JSON following an [Expression](../../cds/cxn) notation. 

> References are either _absolute_ or _relative_. Absolute refs always have fully qualified name of the type in their first segment. Relative references need other absolute reference to which they relate.

Reviewer: ```suggestion
References are either _absolute_ or _relative_. Absolute references always have the fully qualified name of the type in their first segment. Relative references need another absolute reference to which they relate.
```

> Simplest kind of absolute reference is the reference to the entity set, for example, to all books. 

Reviewer: ```suggestion
The simplest kind of absolute reference is the reference to the entity set, for example, to all books. 
```

### Entity References {#entity-refs}

Entity references specify entity sets and can define the target entity set of a [CQL](../../cds/cql) statement or serve as an argument for event handlers.

A reference consists of _segments_ that define the path from the entity's root to a certain part of it. Each segment has the _identifier_ with the name of the entity or an element and an optional filter _predicate_. These predicates might include other references. The references are not bound to the particular model and are not checked against it while they are being built.

References can be either _absolute_ or _relative_. Absolute references always have the fully qualified name of the type in their first segment.

The simplest kind of absolute reference is a reference to the entity set, for example, to all books.

```java
Books_ books = CQL.entity(Books_.class); // {"ref":["sap.capire.bookshop.Books"]}

StructuredTypeRef ref = bookWithId.asRef(); // or CqnStructuredTypeRef which is a cleaner equivalent type
```

The method `asRef()` seals the reference and makes it immutable.

> Relative references do not specify the type in the first segment and are [element references](/java/working-with-cql/query-api#element-refs) most of the time.

Reviewer: ```suggestion
Relative references do not specify the type in the first segment and are [element references](#element-refs) most of the time.
```

```java
CqnStructuredTypeRef toPagesOfTheBook = CQL.entity(Books_.class,
    CQL.to(List.of(refToAuthor.rootSegment())).asRef()).chapters().pages().asRef();
```

References can be copied to produce new references with in-place modification of its segments as follows.

```java
StructuredTypeRef ref = CQL.entity(Books_.class).filter(b -> b.ID().eq("...")).chapters(c -> c.ID().eq("...")).pages(p -> p.ID().eq("...")).asRef();
RefBuilder<StructuredTypeRef> builder = CQL.copy(ref);
builder.segments().forEach(s -> {
  // add new predicate to each segment
  s.filter(CQL.and(CQL.get(Books.GENRE).eq("Fiction"), s.filter().orElse(null)));
});

StructuredTypeRef copy = builder.build(); // new reference is ready
```

Use the [`CqnAnalyzer`](/java/working-with-cql/query-introspection#cqnanalyzer) to bind the reference back to the CDS model and inspect annotations or filter values.

Use [`CqnVisitor`](/java/working-with-cql/query-introspection#cqnvisitor) to introspect references. 

:::warning Limitation
The references are not comparable between each other. They cannot be used as map keys or set values.
:::

### Element References {#element-refs}

> An element reference points to a regular element of the entity. Such references are usually _relative_ and form select list for a CQL statement or an expand.

Reviewer: ```suggestion
An element reference points to a regular element of the entity. Such references are usually _relative_ and form the select list for a CQL statement or an expand.
```

You can also get [entity references](query-execution#entity-refs) from the result of a CDS QL statement to address an entity via its key values in other statements.

Each reference has ordered sequence of _segments_ that define the path from the entity's root to the certain part of it. Segment has the _identifier_ with the name of the entity or an element and optional filter _predicate_.

Existing reference can be reused as an object or a variable, or a new reference can be built on top of it. References are not bound to the particular model and are not checked against it while they are being built.

References can be _absolute_ or _relative_. Absolute reference has fully qualified entity name as the identifier in the first segment. They usually have associations as their segments.

You start with the reference pointing to a book with certain key. You build it using corresponding [model interfaces](../cqn-services/persistence-services#model-interfaces) providing you the methods that corresponds to the elements of the book.

```java

> StructuredType<?> bookWithId = CQL.entity(Books_.class).filter(b -> b.ID().eq("...")); // Books(ID=...)

Author: What makes the ref structured?

```

To navigate to the author of the book, use `CQL.entity(...)` to make it typed again and add one more segment to it. Note, that this does not check that the original reference is indeed the reference to the book. This only lets you use required model interface.

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]},"author"]}
CqnStructuredTypeRef refToAuthor = CQL.entity(Books_.class, ref).author().asRef(); // [!code focus]
```

With `CQL.to(...)` the same is produced dynamically.

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]},"author"]}
CqnStructuredTypeRef toAuthor = CQL.to(ref.segments()).to("author").asRef(); // [!code focus]
```

This new reference might be used as the root of its own statement to read or change the author.

To navigate to the parent, you can strip segment from the end of the ref.

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]}]}
CqnStructuredTypeRef toParent = CQL.to(
    refToAuthor.segments().subList(0, refToAuthor.segments().size() - 1)).asRef(); // [!code focus]
```

> Method `rootSegment()` that can be used to construct the reference starting with the root of the existing reference. For example, when you need to navigate to different parts of the same structured type, for example, from the author of the book to its pages.

Reviewer: ```suggestion
The `rootSegment()` method can be used to construct the reference starting with the root of the existing reference. For example, when you need to navigate to different parts of the same structured type, for example, from the author of the book to its pages.
```

```java
CqnStructuredTypeRef toPagesOfTheBook = CQL.to(List.of(refToAuthor.rootSegment())).to("chapters").to("pages").asRef();
```

You can use the `CQL.entity(...)` to use the convenience of model interfaces for this with additional casts.

```java
CqnStructuredTypeRef toPagesOfTheBook = CQL.entity(Books_.class,
    CQL.to(List.of(refToAuthor.rootSegment())).asRef()).chapters().pages().asRef();
```

References can be copied to produce new references with in-place modification of its segments as follows.

```java
StructuredTypeRef ref = CQL.entity(Books_.class).filter(b -> b.ID().eq("...")).chapters(c -> c.ID().eq("...")).pages(p -> p.ID().eq("...")).asRef();
RefBuilder<StructuredTypeRef> builder = CQL.copy(ref);
builder.segments().forEach(s -> {
  // add new predicate to each segment
  s.filter(CQL.and(CQL.get(Books.GENRE).eq("Fiction"), s.filter().orElse(null)));
});

StructuredTypeRef copy = builder.build(); // new reference is ready
```

> Test this code thoroughly and ensure that you do not omit the filters or change the reference so that it becomes inconsistent.

Reviewer: ```suggestion
Ensure thorough testing when modifying references to avoid omitting filters or creating inconsistent reference states.
```

Select.from(authors).columns("name"); // [!code focus]
```

The resulting statement has two references: one absolute describing the author and the one relative, describing the name of the author.

References can be analyzed with the [`CqnAnalyzer`](/java/working-with-cql/query-introspection#cqnanalyzer) to bind it back to the model, for example, to find annotations or extract filter values.

> The existing references obtained from the statements or [injected in custom handlers](/java/event-handlers/#entity-reference-arguments) can be used to produce new references. 

Reviewer: Useless, no diff

Given that there is simple reference pointing to the book created as follows.

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]}]}
CqnStructuredTypeRef ref = CQL.entity(Books_.class).filter(b -> b.ID().eq("...")).asRef(); // [!code focus]
```

> To navigate to author of the book, use `CQL.entity(...)` to make it typed again and add one more segment to it. Note, that this does not check that original reference is indeed the reference to the book, this only lets you use required model interface.

Reviewer: ```suggestion
To navigate to the author of the book, use `CQL.entity(...)` to make it typed again and add one more segment to it. Note, that this does not check that the original reference is indeed the reference to the book. This only lets you use required model interface.
```

```suggestion
To navigate to the author of the book, use `CQL.entity(...)` to make it typed again and add one more segment to it. Note, that this does not check that the original reference is indeed the reference to the book. Instead, `CQL.entity(...)` only lets you use required model interface.
```

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]},"author"]}
CqnStructuredTypeRef refToAuthor = CQL.entity(Books_.class, ref).author().asRef(); // [!code focus]
```

With `CQL.to(...)` the same is produced dynamically.

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]},"author"]}
CqnStructuredTypeRef toAuthor = CQL.to(ref.segments()).to("author").asRef(); // [!code focus]
```

This new reference might be used as the root of its own statement to read or change the author.

> With this one, one might need to navigate to the parent again. This is done by stripping the segments from the back of the reference.

Reviewer: Useless, no diff

```java
// {"ref":[{"id":"sap.capire.bookshop.Books","where":[{"ref":["ID"]},"=",{"val":"..."}]}]}
CqnStructuredTypeRef toParent = CQL.to(
    refToAuthor.segments().subList(0, refToAuthor.segments().size() - 1)).asRef(); // [!code focus]
```

There is the method `rootSegment()` that can be used to construct the reference starting with the same root as the existing reference. This is useful when you require navigation to different part of the same structured type, e.g. for example from the book's author to the pages of it.

```java
CqnStructuredTypeRef toPagesOfTheBook = CQL.to(List.of(refToAuthor.rootSegment())).to("chapters").to("pages").asRef();
```

You can use the `CQL.entity(...)` to use convenience of model interfaces for this with additional cast.

```java
CqnStructuredTypeRef toPagesOfTheBook = CQL.entity(Books_.class,
    CQL.to(List.of(refToAuthor.rootSegment())).asRef()).chapters().pages().asRef();
```

> Same references is produced as the result of last two statements. You, of course, have to provide the filters, if to-many associations used in them.

Reviewer: ```suggestion
The same reference is produced as the result of last two statements. You, of course, have to provide the filters, if to-many associations are used in them.
```

References can be copied to produce other refs with in-place modification available for its segments as follows.

```java
StructuredTypeRef ref = CQL.entity(Books_.class).filter(b -> b.ID().eq("...")).chapters(c -> c.ID().eq("...")).pages(p -> p.ID().eq("...")).asRef();
RefBuilder<StructuredTypeRef> builder = CQL.copy(ref);
builder.segments().forEach(s -> {
  // add new predicate to each segment
  s.filter(CQL.and(CQL.get(Books.GENRE).eq("Fiction"), s.filter().orElse(null)));
});

StructuredTypeRef copy = builder.build(); // new reference is ready
```

The code that manipulates the references must be tested thoroughly and you always need to ensure that you do not loose the filters or change the reference so that it becomes inconsistent.

:::warning Limitation
The references are not comparable between each other. They cannot be used as keys of maps or values in a set.
:::

### Elements References {#element-refs}

> Element reference points to an element of the entity. Such references are usually _relative_, they do not have the name of the entity in their root. They also can include filters in their segments except _the last one_. 

Reviewer: ```suggestion
An element reference points to an element of the entity. Such references are usually _relative_, they do not have the name of the entity in their root. They also can include filters in their segments except _the last one_. 
```
